### DISCUSSION TITLE
How to track DSPy call within Langgraph

### DISCUSSION THREAD

**Discussion by marcklingen** [langfuse/langfuse#3553] [Support]
I've just chatted with a team struggling with logging DSPy calls with a langgraph application. The DSPy integration under the hood uses the openai integration which leads to the problem that each dspy call is logged as a separate trace instead of being consolidated in the langgraph trace.

@hassiebp is there a good option via custom run_ids or context to group these together?

**Comment by hassiebp** [comment]
Yes, this is be possible via: 

-> creating a custom trace_id
```python
from uuid import uuid4
trace_id = uuid4()
```
-> passing the `trace_id` as `run_id` to the graph invocation like this
```python
graph.invoke(config={"run_id": trace_id})
```

-> When creating the dspy OpenAI client, pass the `trace_id` here too:
```python
openai = dspy.OpenAI(model='gpt-4o-mini', temperature=0.1, max_tokens=1500, trace_id=trace_id)

openai("Explain DSPy and Langfuse to me")
```

Let me know if I should share a full example 👍🏾

**Reply by marcklingen** [reply]
Is this also possible on a run_id level in a nested span/graph-node? Afaik run_ids can only be set on the topmost trace level

**Reply by hassiebp** [reply]
Unfortunately not possible with the current integration. The DSpy calls will be grouped directly under the trace with the abovementioned workaround
